app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,156 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.home.blocklist
+
+import androidx.annotation.VisibleForTesting
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.support.ktx.kotlin.sha1
+import org.mozilla.fenix.home.HomeFragmentAction
+import org.mozilla.fenix.home.HomeFragmentState
+import org.mozilla.fenix.home.recentbookmarks.RecentBookmark
+import org.mozilla.fenix.home.recenttabs.RecentTab
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem
+import org.mozilla.fenix.utils.Settings
+
+/**
+ * This [Middleware] reacts to item removals from the home screen, adding them to a blocklist.
+ * Additionally, it reacts to state changes in order to filter them by the blocklist.
+ *
+ * @param settings Blocklist is stored here as a string set
+ */
+class BlocklistMiddleware(
+    private val settings: Settings
+) : Middleware<HomeFragmentState, HomeFragmentAction> {
+
+    private var blocklist: Set<String> = settings.homescreenBlocklist
```

Comment:
This can be a `val` currently because you're not writing to the `blocklist` field. However, this actually points to a problem. The middleware (and therefore this field) is initialized when the home fragment is created pointing to the instance of `settings.homescreenBlocklist` at that time. Below when handling updates, you're not mutating `homescreenBlocklist`, but setting it to a new instance `updatedBlockList` e.g.: https://github.com/mozilla-mobile/fenix/pull/23070/files#diff-4c994d58d9a22e516dcec6ef8d81c5252ccbcdbbec011d32d12b532189b2fcf2R82

So the `blocklist` field of the middelware will be out-of-sync with `settings` after the update. You could make this work by turning this into a `getter` functino so the lookup happens every time, but I recommend simplifying and adding `val blocklist = settings.homescreenBlocklist` to `invoke` directly.

Reply:
The implementation that this touched has moved to `BlocklistHandler`, but should avoid this issue now.